Prototope/Behaviors.swift

```diff
@@ -0,0 +1,194 @@
+    }
+}
+
+//MARK: Behavior Bindings
+
+// Abstract. Describes a relationship between a layer and a single instance of a behavior.
+// Subclasses should encapsulate the state necessary to handle the behavior correctly.
+class BehaviorBinding: Equatable, Hashable {
+    let id: Int
+    let hostLayer: Layer
+    
+    static var behaviorCounter = 0
+    
+    init(hostLayer: Layer) {
+        self.id = BehaviorBinding.behaviorCounter++
+        self.hostLayer = hostLayer
+    }
+    
+    func update() {
+        fatalError("BehaviorBinding.update must be overridden")
+    }
+    
+    var hashValue: Int {
+        return id
+    }
+}
+
+func ==(b1: BehaviorBinding, b2: BehaviorBinding) -> Bool {
+    return b1.id == b2.id
+}
+
+// Possible collision states for a pair of layers
+enum CollisionState {
+    case NonOverlapping
+    case PartiallyIntersects
+    case ContainedIn
+    case Contains
+    
+    static func stateForLayer(layer1: Layer, andLayer layer2: Layer) -> CollisionState {
+        let rect1 = CGRect(layer1.frame)
+        let rect2 = CGRect(layer2.frame)
+        
+        if !CGRectIntersectsRect(rect1, rect2) {
+            return .NonOverlapping
+        }
+        
+        if CGRectContainsRect(rect1, rect2) {
+            return .Contains
+        } else if CGRectContainsRect(rect2, rect1) {
+            return .ContainedIn
+        }
+        
+        return .PartiallyIntersects
+    }
+}
+
+// Concrete class encapsulating the necessary state for CollisionBehaviors
+class CollisionBehaviorBinding : BehaviorBinding {
+    var previousState: CollisionState
+    let config: CollisionBehavior
+    
+    init(hostLayer: Layer, config: CollisionBehavior) {
+        self.config = config
+        self.previousState = CollisionState.stateForLayer(hostLayer, andLayer: self.config.otherLayer)
+        
+        super.init(hostLayer: hostLayer)
+    }
+
+    override func update() {
+        self.updateWithState(CollisionState.stateForLayer(self.hostLayer, andLayer: self.config.otherLayer))
+    }
+    
+    func updateWithState(state: CollisionState) {
+        let old = previousState
+        switch (self.config.kind, old, state) {
+            
+        case (.Entering,.NonOverlapping,.PartiallyIntersects):
+            fallthrough
+        case (.Entering,.NonOverlapping,.ContainedIn):
+            fallthrough
+        case (.Entering,.NonOverlapping,.Contains):
+            fire()
+            
+        case (.Leaving,.PartiallyIntersects,.NonOverlapping):
+            fallthrough
+        case (.Leaving,.ContainedIn,.NonOverlapping):
+            fallthrough
+        case (.Leaving,.Contains,.NonOverlapping):
+            fire()
+            
+        default:
+            ()
+        }
+        
+        self.previousState = state
+    }
+    
+    func fire() {
+        self.config.handler()
+    }
+}
+
+//MARK: Global Behavior Store
+
+// Global bag of layer behavior state
+class BehaviorStore {
```

Comment:
`Store` sounds like a value to me, but this thing behaves—it manages the heartbeat and owns control flow into the behaviors.

Perhaps `BehaviorDriver`?